File: protocol/MessageInfo.cs
using System.Globalization;

namespace protocol
{
    class MessageInfo
    {
        public string SignalType { get; set; }
        public string MessageType { get; set; }
        public double Latitude { get; set; }
        public double Longitude { get; set; }
        public string Message { get; set; }

        public static MessageInfo Build(string message)
        {
            try
            {
                var messageInfo = new MessageInfo();
                string[] parts = message.Split('#');

                if (parts.Length != 6)
                {
                    throw new FormatException("incorrect nr of segments");
                }

                if (!parts[parts.Length - 1].Equals("END"))
                {
                    throw new FormatException("does not end in END");
                }

                messageInfo.SignalType = parts[0];
                messageInfo.MessageType = parts[1];
                messageInfo.Latitude = double.Parse(parts[2], CultureInfo.InvariantCulture); // keeps the dot as decimal thing
                messageInfo.Longitude = double.Parse(parts[3], CultureInfo.InvariantCulture); 

                messageInfo.Message = parts[4];

                return messageInfo;
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error processing message: {ex.Message}");
            }

            return null;
        }

        public string BuildResponse()
        {
            var responseMessage = GetResponseMessage();
            return $"ACK#MSG#123.4567#123.4567#{responseMessage}#END";
        }

        string GetResponseMessage()
        {
            switch (SignalType)
            {
                case "EMG":
                    switch (MessageType)
                    {
                        case "SOS":
                            return "Dispatching rescue crews";
                        case "PAN":
                            return "Dispatching rapid-response team";
                        case "FIR":
                            return "Dispatching firefighting vessel";
                        case "INT":
                            return "Tracking unidentified craft";
                        case "MED":
                            return "Dispatching medical team";
                        default:
                            return "/";
                    }
                case "MNT":
                    switch (MessageType)
                    {
                        case "CHK":
                            return "Dispatching inspection crew";
                        case "REP":
                            return "Dispatching repair crew";
                        case "REF":
                            return "Dispatching tanker vessel";
                        case "CLN":
                            return "Dispatching specialized cleaning crew";
                        default:
                            return "/";
                    }
                case "INF":
                    return "Asteroid trajectory noted";
                case "REQ":
                    return "Dispatching emergency-supply vessel";
                default:
                    return "/";
            }
        }
    }
}

File: GOLANG/protocol/messageinfo.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type MessageInfo struct {
	SignalType  string
	MessageType string
	Latitude    float64
	Longitude   float64
	Message     string
}

func (m *MessageInfo) String() string {
	return fmt.Sprintf("SignalType: %s\nMessageType: %s\nLatitude: %f\nLongitude: %f\nMessage: %s",
		m.SignalType, m.MessageType, m.Latitude, m.Longitude, m.Message)
}

func BuildMessageInfo(message string) (*MessageInfo, error) {
	parts := strings.Split(message, "#")

	if len(parts) != 6 {
		return nil, fmt.Errorf("invalid message format: incorrect number of segments")
	}

	if parts[len(parts)-1] != "END" {
		return nil, fmt.Errorf("invalid message format: does not end in END")
	}

	latitude, err := strconv.ParseFloat(parts[2], 64)
	if err != nil {
		return nil, fmt.Errorf("error parsing latitude: %v", err)
	}

	longitude, err := strconv.ParseFloat(parts[3], 64)
	if err != nil {
		return nil, fmt.Errorf("error parsing longitude: %v", err)
	}

	return &MessageInfo{
		SignalType:  parts[0],
		MessageType: parts[1],
		Latitude:    latitude,
		Longitude:   longitude,
		Message:     parts[4],
	}, nil
}

func (m *MessageInfo) Serialize() string {
	signalMeaning := m.getSignalMeaning()
	return fmt.Sprintf("ACK#MSG#1%s#1%s#%s#END",
		doubleToString(m.Latitude),
		doubleToString(m.Longitude),
		signalMeaning)
}

func doubleToString(value float64) string {
	return strconv.FormatFloat(value, 'f', -1, 64)
}

func (m *MessageInfo) getSignalMeaning() string {
	switch m.SignalType {
	case "EMG":
		switch m.MessageType {
		case "SOS":
			return "Dispatching rescue crews"
		case "PAN":
			return "Dispatching rapid-response team"
		case "FIR":
			return "Dispatching firefighting vessel"
		case "INT":
			return "Tracking unidentified craft"
		case "MED":
			return "Dispatching medical team"
		default:
			return "/"
		}
	case "MNT":
		switch m.MessageType {
		case "CHK":
			return "Dispatching inspection crew"
		case "REP":
			return "Dispatching repair crew"
		case "REF":
			return "Dispatching tanker vessel"
		case "CLN":
			return "Dispatching specialized cleaning crew"
		default:
			return "Maintenance: Unknown type"
		}
	case "INF":
		return "Asteroid trajectory noted"
	case "REQ":
		return "Dispatching emergency-supply vessel"
	default:
		return "Unknown message"
	}
}
